Adding Multi User & Role Tipping and Help/Warning Optimisation

ProfessorDey opened the pull request:
This includes my previous pull request. See individual commits for details

Commit:
Optimise Help Msg & Wrong Channel Warning

This one involves a fair amount of reworking of the way functions operate, all checks for being in private or bot sandbox have been removed from functions and that functionality is now applied through privateOrSandboxOnly(), passing the message, channel warning, function to be called and then an array of arguments, sans message, as that's automatically attached at the head as all primary functions require the message anyway. while this makes each case entry a bit longer, it does reduce code reuse and potential for old code being left in, since each function had to previously be updated individually if something happened with the way these tests were carried out.

I've yet to find a way to allow us to set the export description and the help message in one fell swoop, so for now I've manually set the description. The help message itself is now broken down into pairs of subcommand and description, the subcommand will be indented once, while the description will be indented twice.

The way the help function now works is to allow everything to be set at the start, without needing to hunt all the way down to find the duplicate help message. Now everything can be replaced in the main function, though this will require a bit of computation with each request presently, as I don't know if or how the exported functions can access constants defined outside them. This can be optimised later if need be and at the very least shouldn't slow things down very much with the few entries that there are.

Unfortunately changing the way that the help message works does involve parameterising the help message itself for any function that requires it, but this also provides a helpful indicator of what functions might fail and fall back on the help function.

Commit:
Merge pull request #1 from ProfessorDey/patch-4

Update to latest

Commit:
Adding Multi User Tipping and Role Tipping

This is actually not too substantial an addition. it adds 2 additional command exports for tipbot, namely !multitip and !roletip.

!multitip [private] <user>+ <amount>
The more complex new feature, this command will take a list of users, seperated by spaces, and passes through to doTip() if you only list a single user. It tests each word past the ! command and private tag, checking for a user mention using the regex.test() function of the Discord.js USERS_PATTERN, using that to count the number of users mentioned in a row, then takes the first word that is not a match to be the amount. Error checking then returns the appropriate errors to the user, if any. Otherwise it proceeds to send tips to each user individually. this results in messages for every user that receives a tip, which can be messy. It will also message the author once for every user a tip is sent to, if using private mode, this isn't ideal, but would require rewriting the sendLbc() function, which is outside the scope of this commit.

!roletip <role> <amount>
This is a relatively simpler feature, the command taking a single role and an amount, then extracting the userIDs from the role via the Roles.members and GuildMember.user values. It will return seperate errors for a lack of a role in the message and the lack of any users in a role. Like the !multitip command, it will send tips individually to each user, with the same spam of messages either in the channel or the author's private channel.

aayanl commented on bot/modules/tipbot.js:
```diff
@@ -111,11 +166,86 @@ function doTip(message, tipper, words) {
+  }
+  let amount = getValidatedAmount(words[amountOffset]);
+  if (amount == null) {
+    message.reply('I dont know how to tip that many credits');
```
don't*

aayanl commented on bot/modules/tipbot.js:
````diff
@@ -6,25 +6,88 @@ config = config.get('lbrycrd');
+                      ['withdraw ADDRESS AMOUNT', 'Withdraw AMOUNT credits to ADDRESS'],
+                      ['[private] <user> <amount>', 'Mention a user with @ and then the amount to tip them, or put private before the user to tip them privately.']],
+      helpmsg = '```**!tip**\n' + formatDescriptions(helpmsgparts) + 'Key: [] : Optionally include contained keyword, <> : Replace with appropriate value.```',
+      channelwarning = 'Please use <#369896313082478594> or DMs to talk to bots.';
````
The channel shouldn't be hard-coded, rather in a config value somewhere

aayanl commented on bot/modules/tipbot.js:
````diff
@@ -6,25 +6,88 @@ config = config.get('lbrycrd');
+                        ['<user>+ <amount>', 'Mention one or more users in a row, seperated by spaces, then an amount that each mentioned user will receive.'],
+                        ['private <user>+ <amount>','Put private before the user list to have each user tipped privately, without revealing other users tipped.']],
+        helpmsg = '```**!multitip**\n' + formatDescriptions(helpmsgparts) + 'Key: [] : Optionally include contained keyword, <> : Replace with the appropriate value, + : Value can be repeated for multiple entries.```',
+        channelwarning = 'Please use <#369896313082478594> or DMs to talk to bots.';
````
ditto

WaveringAna committed:
Grammar fixes and un-hardcode bot-sandbox

Commit:
Grammar fixes and un-hardcode bot-sandbox

filipnyquist commented:
Looks good, i´ll test run this tomorrow on my sandbox server, i´ll assign myself this issue as nikoo catches up with life after studies :)

Commit:
Implement MSFTblockchain's Fixes

Implements MSFTblockchain's fixes from PR #4 without customisations or version regression.

Now requires Node.js Regex module for testing for user mentions in text with regex, including a fallback if a unit test fails.

Also adds !tipcommands command which gives an up to date list of all commands tipbot accepts with a brief description of each.

The tip message tail about DMing tipbot is now contextualised to give the appropriate !<command> and also points users towards the new !tipcommands command.

Private tip messages are now a bit nicer to read and prv is now a boolean flag to fix some odd errors that were occuring.

Error Messages now delete themselves after a short time for cleanliness.

Help messages are now in Discord Embeds to line up with Wunderbot, though this currently ignores tabs, so help messages don't quite match the requested format. Commands have been made bold to prevent confusion.

Everything should now actually work, but will require testing to be sure.

ProfessorDey commented:
Code has been updated to include some code fixes from MSFTblockchain and also adds some additional functionality, while cleaning things up code wise.

Commit:
Final functionality fixes from MSFTblockchain

All finally operational and tested on MSFT's RVN server, code ported back over after figuring out various JS specific implementation issues. Now all works correctly and doesn't include anyone it shouldn't.

ProfessorDey commented:
And we are finally fully operational! Thanks a lot MSFTblockchain, you will be receiving half of whatever bounty or reward (personal or otherwise, Brazen likes his new custom features).

MSFTserver commented:
roletip if user doesnt have an address assigned it fails to assign one and spams error 500

ProfessorDey commented:
Unfortunately have no idea about the operation of the lbry backend because can't find any of that code anywhere. Also have no idea how new users are added as it's apparently supposed to be automatic when tipbot sends to a new address, meaning it _should_ be in lbry.sendFrom(), which I don't have access to.

MSFTserver commented:
well this bot uses the bitcoind wallets and rpc commands if that helps

MSFTserver commented:
NOTE: if you encounter some errors with sending transaction to many people
i.e `Error: Invalid params, response status code: 500` 
if you check the debug log in your wallet folder youll notice lines say: 
`request rejected becasue http work queue defph exceeded, it can be increased with the -rpcworkqueue= setting`
simply add this to the wallet config 

| command | | description |
|:----------|-|:----------|
| rpcthreads=X | | Set the number of threads to service RPC calls (default: 4) |
| rpcworkqueue=XXX | | Sets the depth of the work queue to service RPC calls (default is 16) |
| rpcclienttimeout=X | | How many seconds bitcoin will wait for a complete RPC HTTP request |

MSFTserver commented:
roletip and multi private tipping mode isnt working

nikooo777 commented:
as discussed with @ProfessorDey there are still a few things to be committed before this can be merged so I'll wait for him to give me to GO.

Commit:
Integrating final MSFT fixes

Now the last little changes have been implemented, had to modify how we got the user to message since we were previously getting the first user mentioned in the message, whereas we now match against all known users in the guild. This means users actually get messaged by, say, rolebot even if they're not in the channel you're tipping from. Of course you can't directly mention anyone not in the channel for multitip, so that's a non-issue in most use cases.

Also had to ensure certain function calls were sending the right type, because dynamic languages and lack of type safety.

That _should_ be absolutely everything needed to have it fully functional.

ProfessorDey commented:
Ok @nikooo777 we're finally done, that should now be fully functional so long as you remember to install regex and set the rpc queue settings to higher values, since we have some large roles after all.

nikooo777 commented:
ok my changes have created conflicts. I'll see what i can do about it

nikooo777 committed:
Merge branch 'master' into multirole

ported over the changes made in master
changed vars to let
simplified code
added missing dependency to package.json

nikooo777 committed:
updated default configuration

nikooo777 commented:
Alright It took me a while but I think i managed to clean it up. Pushed to your fork so you can now test it again before I merge it.

nikooo777 committed:
missed 2 coins -> credits

ProfessorDey commented:
As I've said, I can confirm this works fine on my end, so feel free to merge whenever.

nikooo777 commented:
Thank you for your contribution! /merged
